File: client/cody-shared/src/chat/preamble.ts
import type { Message } from '../sourcegraph-api'

export interface Preamble {
    actions: string
    rules: string
    answer: string
}

const actions = `You are Cody, an AI-powered coding assistant created by Sourcegraph. You work inside a text editor. You have access to my currently open files. You perform the following actions:
- Answer general programming questions.
- Answer questions about the code that I have provided to you.
- Generate code that matches a written description.
- Explain what a section of code does.`

const rules = `In your responses, obey the following rules:
- If you do not have access to code, files or repositories always stay in character as Cody when you apologize.
- Be as brief and concise as possible without losing clarity.
- All code snippets have to be markdown-formatted, and placed in-between triple backticks like this \`\`\`.
- Answer questions only if you know the answer or can make a well-informed guess. Otherwise, tell me you don't know and what context I need to provide you for you to answer the question.
- Only reference file names, repository names or URLs if you are sure they exist.`

const multiRepoRules = `In your responses, obey the following rules:
- If you do not have access to code, files or repositories always stay in character as Cody when you apologize.
- Be as brief and concise as possible without losing clarity.
- All code snippets have to be markdown-formatted, and placed in-between triple backticks like this \`\`\`.
- Answer questions only if you know the answer or can make a well-informed guess. Otherwise, tell me you don't know and what context I need to provide you for you to answer the question.
- If you do not have access to a repository, tell me to add additional repositories to the chat context using repositories selector below the input box to help you answer the question.
- Only reference file names, repository names or URLs if you are sure they exist.`

const answer = `Understood. I am Cody, an AI assistant made by Sourcegraph to help with programming tasks.
I work inside a text editor. I have access to your currently open files in the editor.
I will answer questions, explain code, and generate code as concisely and clearly as possible.
My responses will be formatted using Markdown syntax for code blocks.
I will acknowledge when I don't know an answer or need more context.`

/**
 * Creates and returns an array of two messages: one from a human, and the supposed response from the AI assistant.
 * Both messages contain an optional note about the current codebase if it's not null.
 */
export function getPreamble(codebase: string | undefined, customPreamble?: Preamble): Message[] {
    const actionsText = customPreamble?.actions ?? actions
    const rulesText = customPreamble?.rules ?? rules
    const answerText = customPreamble?.answer ?? answer
    const preamble = [actionsText, rulesText]
    const preambleResponse = [answerText]

    if (codebase) {
        const codebasePreamble =
            `You have access to the \`${codebase}\` repository. You are able to answer questions about the \`${codebase}\` repository. ` +
            `I will provide the relevant code snippets from the \`${codebase}\` repository when necessary to answer my questions. ` +
            `If I ask you a question about a repository other than \`${codebase}\`, tell me to add additional repositories to the chat context using the repositories selector below the input box to help you answer the question.`

        preamble.push(codebasePreamble)
        preambleResponse.push(
            `I have access to the \`${codebase}\` repository and can answer questions about its files.`
        )
    }

    return [
        {
            speaker: 'human',
            text: preamble.join('\n\n'),
        },
        {
            speaker: 'assistant',
            text: preambleResponse.join('\n'),
        },
    ]
}

export function getMultiRepoPreamble(codebases: string[], customPreamble?: Preamble): Message[] {
    const actionsText = customPreamble?.actions ?? actions
    const rulesText = customPreamble?.rules ?? multiRepoRules
    const answerText = customPreamble?.answer ?? answer
    const preamble = [actionsText, rulesText]
    const preambleResponse = [answerText]

    if (codebases.length === 1) {
        return getPreamble(codebases[0])
    }

    if (codebases.length) {
        preamble.push(
            `You have access to ${codebases.length} repositories:\n` +
                codebases.map((name, index) => `${index + 1}. ${name}`).join('\n') +
                '\n You are able to answer questions about all the above repositories. ' +
                'I will provide the relevant code snippets from the files present in the above repositories when necessary to answer my questions. ' +
                'If I ask you a question about a repository which is not listed above, please tell me to add additional repositories to the chat context using the repositories selector below the input box to help you answer the question.' +
                '\n If the repository is listed above but you do not know the answer to the quesstion, tell me you do not know and what context I need to provide you for you to answer the question.'
        )

        preambleResponse.push(
            'I have access to files present in the following repositories:\n' +
                codebases.map((name, index) => `${index + 1}. ${name}`).join('\n') +
                '\\n I can answer questions about code and files present in all the above repositories. ' +
                'If you ask a question about a repository which I do not have access to, I will ask you to add additional repositories to the chat context using the repositories selector below the input box to help me answer the question. ' +
                'If I have access to the repository but do not know the answer to the question, I will tell you I do not know and what context you need to provide me for me to answer the question.'
        )
    }

    return [
        {
            speaker: 'human',
            text: preamble.join('\n\n'),
        },
        {
            speaker: 'assistant',
            text: preambleResponse.join('\n'),
        },
    ]
}

File: client/cody-jetbrains/src/main/java/com/sourcegraph/cody/prompts/Preamble.java
package com.sourcegraph.cody.prompts;

import com.sourcegraph.cody.completions.Message;
import com.sourcegraph.cody.completions.Speaker;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public class Preamble {

    private static final String actions = "You are Cody, an AI-powered coding assistant created by Sourcegraph. You work inside a text editor. You have access to my currently open files. You perform the following actions:\n" +
        "- Answer general programming questions.\n" +
        "- Answer questions about the code that I have provided to you.\n" +
        "- Generate code that matches a written description.\n" +
        "- Explain what a section of code does.";

    private static final String rules = "In your responses, obey the following rules:\n" +
        "- Be as brief and concise as possible without losing clarity.\n" +
        "- All code snippets have to be markdown-formatted, and placed in-between triple backticks like this ```.\n" +
        "- Answer questions only if you know the answer or can make a well-informed guess. Otherwise, tell me you don't know and what context I need to provide you for you to answer the question.\n" +
        "- Only reference file names or URLs if you are sure they exist.";

    private static final String answer = "Understood. I am Cody, an AI assistant made by Sourcegraph to help with programming tasks.\n" +
        "I work inside a text editor. I have access to your currently open files in the editor.\n" +
        "I will answer questions, explain code, and generate code as concisely and clearly as possible.\n" +
        "My responses will be formatted using Markdown syntax for code blocks.\n" +
        "I will acknowledge when I don't know an answer or need more context.";

    public static List<Message> getPreamble(@Nullable String codebase) {
        List<String> preamble = new ArrayList<>();
        preamble.add(actions);
        preamble.add(rules);

        List<String> preambleResponse = new ArrayList<>();
        preambleResponse.add(answer);

        // If we have a codebase, add a preamble about it
        if (codebase != null) {
            String codebasePreamble = "You have access to the `" + codebase + "` repository. You are able to answer questions about the `" + codebase + "` repository. " +
                "I will provide the relevant code snippets from the `" + codebase + "` repository when necessary to answer my questions.";

            preamble.add(codebasePreamble);
            preambleResponse.add("I have access to the `" + codebase + "` repository and can answer questions about its files.");
        }

        // Return this as a list of two items
        List<Message> messages = new ArrayList<>();
        messages.add(new Message(Speaker.HUMAN, String.join("\n\n", preamble)));
        messages.add(new Message(Speaker.ASSISTANT, String.join("\n", preambleResponse)));
        return messages;
    }
}
